Fix auto fallback for --alsa-mixer-device and --alsa-mixer-index

JasonLG1979 opened the pull request:
As mentioned in https://github.com/librespot-org/librespot/issues/898#issuecomment-986528998

`--alsa-mixer-device` and `--alsa-mixer-index` now fallback to the card and index specified in `--device`.

roderickvd commented:
Would be nice if @coderootme and yourself could report back on various combinations, also in the long form of `hdmi:CARD=HDMI,DEV=0`.

coderootme commented:
Yes, I can test it no problem!

JasonLG1979 commented:
> Would be nice if @coderootme and yourself could report back on various combinations, also in the long form of `hdmi:CARD=HDMI,DEV=0`.

It works for me. Where my `--device` is `hw:CARD=Generic_1,DEV=0` or `hw:1,0` I get:

```
[2021-12-17T08:47:44Z INFO  librespot_playback::mixer::alsamixer] Mixing with Alsa and volume control: Log(0.0) for device: hw:CARD=Generic_1 with mixer control: PCM,0
```
And:

```
[2021-12-17T08:43:14Z INFO  librespot_playback::mixer::alsamixer] Mixing with Alsa and volume control: Log(0.0) for device: hw:1 with mixer control: PCM,0
```

Everything works as intended.

Note that `hw:CARD=Generic_1,DEV=0` and `hw:1,0` are the same exact output just different ways to express them. It's preferable though to use the `hw:CARD=Generic_1,DEV=0` form as it's possible for the index of a card to change boot to boot, so it's more stable than `hw:1,0`.

JasonLG1979 commented:
After this quick fix is merged I will go though and de-panic the mixer(s).

JasonLG1979 commented:
As a future improvement `--alsa-mixer-index` could also auto fallback to the dev index specified in `--device` if at all. So in that case for example a `--device` of either `hw:CARD=Generic_1,DEV=1` or `hw:1,1` becomes the equivalent of `--alsa-mixer-index 1`. None of my devices actually have a mixer on `hw:x,1` so I can't really test it but the logic is sound.

`--alsa-mixer-control` could also default to trying both `PCM` and `Master` if not specified instead of just 'PCM'.

JasonLG1979 commented:
`--device` really should be `--card` or maybe `--audio-output`(?) since `--device` conflicts with ALSA naming conventions where a ALSA `device` is an input or an output on a `card`. I kinda like `--audio-output` since it can be either hardware or software based and `--device` and `--card` both sound like they might be hardware only.

But that's just my opinion.

roderickvd commented:
What do the other backends call it? Particularly `rodio`, PulseAudio and GStreamer as all of those seem to be used pretty commonly.

JasonLG1979 commented:
PulseAudio and Gstreamer both call outputs Sinks. I don't know about rodio?

JasonLG1979 commented:
In Gstreamer an element has an input called a `sink` and an output called a `src` elements connect their `src` to the next element's `sink`. A `sink` element like an audio output has no `src` as it's the end of the audio pipeline. A `src` element has no `sink` since it is the beginning of the audio pipeline.

https://gstreamer.freedesktop.org/documentation/application-development/basics/elements.html?gi-language=c

PulseAudio works about the same way.

https://gavv.github.io/articles/pulseaudio-under-the-hood/

JasonLG1979 commented:
`rodio` seems to work about the same way:

https://docs.rs/rodio/latest/rodio/struct.Sink.html

So basically everyone calls it a sink except ALSA. Which I can honestly understand since they all represent audio signal flow more abstractly while ALSA is lower level. Where all of the different "sink" abstractions will eventually hit the "real" ALSA "card,device".

JasonLG1979 commented:
So I guess `--sink` is a better name than `--device` and the concept of a sink translates across different backends better than device.

JasonLG1979 commented:
@roderickvd that last commit fixes the auto fallback for `--alsa-mixer-index`. So instead of always falling back to 0 if not explicitly set it will fallback to the device index part of "hw:x,x" or "hw:CARD=x,DEV=x" if it's present in `--device`.

JasonLG1979 commented:
Looking for testing from you @coderootme and a review and possibly merge from @roderickvd.

JasonLG1979 commented:
On related side note I looked into unpanicking the mixer and there are 2 things I see as a temporary blocker for me:

1. The mixer(s) touch network code so I'd rather wait for you to finish your work on the new API @roderickvd.

2. Much like the backends `open` can not return an error. Either the mixers will have to be refactored to be able to return an error from `open` or the alsa mixer will need to be refactored to move all panicky code out of `open`.

JasonLG1979 commented:
Well hold up on the merge. auto fallback doesn't work with things like `plughw:0,0` because the mixer must be `hw:`

JasonLG1979 commented:
The last commit should fix fallbacks for devices that don't start with `hw:`. Basically it drops the prefix before the `:` and adds `hw` so for example `plughw:CARD=Generic_1` becomes `hw:CARD=Generic_1` because although `plughw:CARD=Generic_1` is a valid device it is not a valid mixer. The mixer for `plughw:CARD=Generic_1` is `hw:CARD=Generic_1`.

coderootme commented:
> Looking for testing from you @coderootme and a review and possibly merge from you @roderickvd.

Just compiling 042f5f95f787f71fc1a69d17a9a4a82a546b0c20 on a Raspberry Pi 4. Will report back in a few minutes!

coderootme commented:
OK, so here is my test.

I compiled with:
```
$ git clone -b fix-alsa-mixer-device https://github.com/JasonLG1979/librespot.git
$ cd librespot/
$ cargo build --no-default-features --features "alsa-backend" --release
```

coderootme commented:
OK, so here is my test.

I compiled with:

```
$ git clone -b fix-alsa-mixer-device https://github.com/JasonLG1979/librespot.git
$ cd librespot/
$ cargo build --no-default-features --features "alsa-backend" --release
```

My test on a **DAC which has a hardware mixer** &ndash; the one that previously failed to work without ``--alsa-mixer-device`` specified.

#### current librespot 0.3.1 7160dc1:

``/usr/bin/librespot --mixer alsa --device "hw:1,0" --alsa-mixer-device "hw:1"`` -- works properly, as expected.

#### current librespot 0.3.1 7160dc1:

``/usr/bin/librespot --mixer alsa --device "hw:1,0"`` -- panics, as we already know -- could not open alsa mixer.

#### JasonLG1979's fix-alsa-mixer-device, commit 042f5f95f787f71fc1a69d17a9a4a82a546b0c20:

``/usr/bin/librespot_new --mixer alsa --device "hw:1,0" --alsa-mixer-device "hw:1"`` -- still works properly, as we would hope.

#### JasonLG1979's fix-alsa-mixer-device, commit 042f5f95f787f71fc1a69d17a9a4a82a546b0c20:

``/usr/bin/librespot_new --mixer alsa --device "hw:1,0"`` -- now works properly, finding and opening the correct alsa mixer.
``/usr/bin/librespot_new --mixer alsa --device "hw:CARD=Audio,DEV=0"`` -- also works.
``/usr/bin/librespot_new --mixer alsa --device "plughw:CARD=Audio,DEV=0"`` -- also works.

Is this a good enough test for you guys, or could I test something else, or report some more information?

JasonLG1979 commented:
> JasonLG1979's fix-alsa-mixer-device, commit 042f5f9:
/usr/bin/librespot_new --mixer alsa --device "hw:1,0" --alsa-mixer-device "hw:1" -- still works properly, as we would hope.

> JasonLG1979's fix-alsa-mixer-device, commit 042f5f9:
/usr/bin/librespot_new --mixer alsa --device "hw:1,0" -- now works properly, finding and opening the correct alsa mixer.
/usr/bin/librespot_new --mixer alsa --device "hw:CARD=Audio,DEV=0" -- also works.
/usr/bin/librespot_new --mixer alsa --device "plughw:CARD=Audio,DEV=0" -- also works.

Awesome, I also fixed/handled some other situations.

1. While `--device` values like `pulse`, `default`, and `jack` may be valid devices there is no way (that I can tell anyway?) to get the mixer for those "virtual" devices. So if a device is not the `something:x,x` or `something:CARD=x,DEV=x` type `librespot` will exit with an error if `--mixer` is `alsa` and `--alsa-mixer-device` is not specified.

2. All `something:x,x` or `something:CARD=x,DEV=x` type devices are turned to `hw:x,x` or `hw:CARD=x,DEV=x as a fallback when `--alsa-mixer-device` is not specified, because generally as far as I can tell mixers have to be `hw:`. `plughw:CARD=Audio` will not work but `hw:CARD=Audi` will for example.

JasonLG1979 commented:
> Is this a good enough test for you guys, or could I test something else, or report some more information?

I think that about covers it that you.

@roderickvd balls in your court.

JasonLG1979 commented:
@coderootme I guess I forgot:

```
--mixer alsa --device hw:1
--mixer alsa --device hw:CARD=Audio
--mixer alsa --device plughw:CARD=Audio
```

Should also all work in your case can you test those also?

Sorry.

JasonLG1979 commented:
`0` is the default device index. It's implied if you omit it.

roderickvd commented:
> On related side note I looked into unpanicking the mixer and there are 2 things I see as temporary blockers for me:
> 
> 1. The mixer(s) touch network code so I'd rather wait for you to finish your work on the new API @roderickvd.
> 2. Much like the backends `open` can not return an error. Either the mixers will have to be refactored to be able to return an error from `open` or the alsa mixer will need to be refactored to move all panicky code out of `open`.

I think it’s a good idea to have `open` and possibly others to return a `Result` if the sink can fail. But I suggest you wait for me to introduce the global ‘Error` type before you do it. I decided to work on that before moving to other things, and am glad that I did. Refactoring all the `Error` return types is quite a lot of work that I would like to spare you.

JasonLG1979 commented:
> Refactoring all the Error return types is quite a lot of work that I would like to spare you.

Have at it. It's all yours,lol!!!

I think I've done all I can here without changing the mixer(s).

JasonLG1979 commented:
Another question why `open` and not `new` `open` seems weird to me?

coderootme commented:
> ```
> --mixer alsa --device hw:1
> --mixer alsa --device hw:CARD=Audio
> --mixer alsa --device plughw:CARD=Audio
> ```

Just tested these. All 3 work properly.

JasonLG1979 commented:
> Just tested these. All 3 work properly.

Great!!! Thank you very much. @roderickvd what say you?

JasonLG1979 commented:
I can squash when you're ready.

JasonLG1979 commented:
A little force push to get past a code 101.

roderickvd commented on src/main.rs:
```diff
@@ -805,40 +805,113 @@ fn get_setup() -> Setup {
+                        .chars()
+                        .last()
+                        .unwrap_or_default()
+                        .to_digit(10)
```
Converting to base-10 this way is rather explicit 😆 can't we just do `.parse()` or any of the usual idioms?

JasonLG1979 replied:
This way anything but 0-9 is 0. It's shorter than parse u32 unwrap_or_default().

JasonLG1979 replied:
Hold up. I'm still thinking about this one?

roderickvd replied:
Will this work for card numbers >= 10? That might not be so common but while we're going for robustness here...

JasonLG1979 replied:
> Will this work for card numbers >= 10? That might not be so common but while we're going for robustness here...

That's part of what I'm ciphering.

JasonLG1979 replied:
There, I changed it to:
```rust
                .unwrap_or_else(|| match device {
                    Some(ref device_name) if device_name.contains(',') => {
                        let dev = &device_name[device_name.find(',').unwrap_or_default()..];
                        dev[min(dev.find('=').unwrap_or_default() + 1, dev.len())..]
                            .parse::<u32>()
                            .unwrap_or(mixer_default_config.index)
                    }
                    _ => mixer_default_config.index,
                })
```

That should parse up to a dev index of `u32::MAX`.

JasonLG1979 replied:
> It's pretty smart which I mean both in a good and a cautious way. Meaning: this catches a lot of variants in a way that is not entirely obvious. I think it would be wise to document possible values with a little code documentation for future readers.

Done. I've put comments in the code that explain what it does.

So it now reads:

```rust
                .unwrap_or_else(|| match device {
                    // Look for the dev index portion of --device.
                    // Specifically <dev index> when --device is <something>:CARD=<card name>,DEV=<dev index>
                    // or <something>:<card index>,<dev index>.

                    // If --device does not contain a ',' it does not contain a dev index.
                    // In the case that the dev index is omitted it is assumed to be 0 (mixer_default_config.index).
                    // Malformed --device values will also fallback to mixer_default_config.index.
                    Some(ref device_name) if device_name.contains(',') => {
                        // Turn <something>:CARD=<card name>,DEV=<dev index> or <something>:<card index>,<dev index>
                        // into DEV=<dev index> or <dev index>.
                        let dev = &device_name[device_name.find(',').unwrap_or_default()..]
                            .trim_start_matches(',');

                        // Turn DEV=<dev index> into <dev index> (noop if it's already <dev index>)
                        // and then parse <dev index>.
                        // Malformed --device values will fail the parse and fallback to mixer_default_config.index.
                        dev[dev.find('=').unwrap_or_default()..]
                            .trim_start_matches('=')
                            .parse::<u32>()
                            .unwrap_or(mixer_default_config.index)
                    }
                    _ => mixer_default_config.index,
                })
```

And:

```rust
                    Some(ref device_name) => {
                        // Look for the card name or card index portion of --device.
                        // Specifically <card name> when --device is <something>:CARD=<card name>,DEV=<dev index>
                        // or card index when --device is <something>:<card index>,<dev index>.
                        // --device values like `pulse`, `default`, `jack` may be valid but there is no way to
                        // infer automatically what the mixer should be so they fail auto fallback
                        // so --alsa-mixer-device must be manually specified in those situations.
                        let start_index = device_name.find(':').unwrap_or_default();

                        let end_index = match device_name.find(',') {
                            Some(index) if index > start_index => index,
                            _ => device_name.len(),
                        };

                        let card = &device_name[start_index..end_index];

                        if card.starts_with(':') {
                            // mixers are assumed to be hw:CARD=<card name> or hw:<card index>.
                            "hw".to_owned() + card
                        } else {
                            error!(
                                "Could not find an alsa mixer for \"{}\", it must be specified with `--{}` / `-{}`",
                                device_name,
                                ALSA_MIXER_DEVICE,
                                ALSA_MIXER_DEVICE_SHORT
                            );

                            exit(1);
                        }
                    }
```

JasonLG1979 replied:
The assumption during parsing either is that `--device` are not malformed. If `--device` is malformed it would fail when the backend tried to open it anyway.

JasonLG1979 replied:
@roderickvd is that more to your liking?

JasonLG1979 replied:
A lot of it is just taking advantage of the fact that you can use `find` to walk though a string especially if you know what to expect and/or what you want to find for example:

```rust
random_string.find("random chars").unwrap_or_default()
```
The above will either return the left most index of "random chars" or 0 if the string does not include "random chars".

roderickvd commented on src/main.rs:
```diff
@@ -881,10 +954,12 @@ fn get_setup() -> Setup {
+                if is_alsa_mixer {
+                    0.0
+                } else {
+                    VolumeCtrl::DEFAULT_DB_RANGE
```
Though not entirely in the scope of this PR, should I have added an `impl Default for VolumeCtrl` before? So we could do `VolumeCtrl::default().db_range` or something like that here.

JasonLG1979 replied:
All of the config stuff could use a little love but that's a diffrent PR for after you finish the new api stuff.

roderickvd replied:
That's fine.

roderickvd reviewed (commented):
Nice enhancement. As usual a few questions about idiom.

roderickvd commented:
> Another question why `open` and not `new`, `open` seems weird to me?

That's why I said "`open` and possibly others". I can imagine both occurring. `Err` on `new` meaning: this sink cannot be initialized (e.g. device does not exist or is not accessible). `Err` on `open` meaning: this sink or stream cannot be opened at this time (e.g. the sink exists but is currently taken or otherwise unavailable, but may be available at a later time).

JasonLG1979 commented:
> > Another question why `open` and not `new`, `open` seems weird to me?
> 
> That's why I said "`open` and possibly others". I can imagine both occurring. `Err` on `new` meaning: this sink cannot be initialized (e.g. device does not exist or is not accessible). `Err` on `open` meaning: this sink or stream cannot be opened at this time (e.g. the sink exists but is currently taken or otherwise unavailable, but may be available at a later time).

No what I mean is for new to replace open because open creates a new thing in those cases so it's strange to me that the functions are called open.

roderickvd commented:
> > > Another question why `open` and not `new`, `open` seems weird to me?
> > 
> > 
> > That's why I said "`open` and possibly others". I can imagine both occurring. `Err` on `new` meaning: this sink cannot be initialized (e.g. device does not exist or is not accessible). `Err` on `open` meaning: this sink or stream cannot be opened at this time (e.g. the sink exists but is currently taken or otherwise unavailable, but may be available at a later time).
> 
> No what I mean is for new to replace open because open creates a new thing in those cases so it's strange to me that the functions are called open.

Ah I see. I agree that `open` sounds like a transition of state on something that already exists, and not instantiate something new. So things should quack like "new, start, finish` or "create, open, close" but not both.

JasonLG1979 commented:
For the backends it should be new, start, stop. Open in that case is an implementation detail IMHO. A backend might open on new or start depending.

For the mixers it should just be new, as there's no close or even stop.

JasonLG1979 commented:
> So things should quack like "new, start, finish` or "create, open, close" but not both.

I'm not even really a fan of create as really how is it different from new? Although create better than open, new is a very basic, pretty universal programming idiom.

JasonLG1979 commented:
If there are different or alternative way to make a new thing I prefer `new_from_foo` if there are ways to make slightly different kinds or more complete versions of a thing I would prefer something like `new_with_foo`.

JasonLG1979 commented:
So in my example:

`Creds::new()` = new empty creds.
`Creds::new_from_file(file)` = new creds from a file.
`Creds::new_with_username(username)` = new creds with a username.

JasonLG1979 commented:
@roderickvd this is failing on nightly with:
```
thread 'rustc' panicked at 'failed to lookup `SourceFile` in new context', compiler/rustc_query_impl/src/on_disk_cache.rs:500:22
```

It appears to be caused by this (in code quotes so it doesn't add noise to that issue):
```
https://github.com/rust-lang/rust/issues/92163
```

Other than that I think this is ready to merge. I'm pretty sure I've addressed all of your comments?
